<head>
<link rel="stylesheet" type="text/css" href="eppic-static.css">
</head>

<div class="eppic-iframe">

<img class="eppic-iframe-top-img" src="resources/images/eppic-logo.png">
</img>


<h1>
The EPPIC server
</h1>

<p>
EPPIC (Evolutionary Protein-Protein Interface Classifier) aims at classifying 
the interfaces present in protein crystal structures in order to determine whether they are 
biologically relevant or not.
</p>

<p>
After submitting your protein structure as a PDB/mmCIF file or typing a PDB code you will 
get a view of all the protein-protein interfaces present in the crystal. We assess them 
based on one geometrical and two evolutionary criteria:
</p>

<ul>
<li>
<b>geometry</b>: number of core residues, indicating how good the packing in the interface is
</li>
<li>
<b>core-rim score</b>: ratio of evolutionary signal (sequence entropy) of core residues versus 
rim residues
</li>
<li>
<b>core-surface score</b>: a score (z-score like) of sequence entropy of core residues versus 
random samples of other surface residues
</li>
</ul>


<p>
Each of these indicators casts a vote based on predefined score thresholds. The possible calls are: 
</p>

<ul>
<li>
<font color="green">BIO</font>, the interface is biologically relevant
</li>
<li>
<font color="red">XTAL</font>, the interface is only a crystal lattice contact
</li>
<li>
NOPRED, there is not enough information available to make a decision (usually not enough sequence data)
</li>
</ul>
  


<p>
You can see the actual scores for each of the indicators by hovering the mouse pointer 
over the bio/xtal/nopred calls. Warnings may appear on the right side of the interfaces table
marked by the icon <img src="resources/icons/warning_icon.png"></img>. By hovering the mouse over
the icon, one can see different issues for which the prediction should be taken with care, e.g.
engineered residues at the core or rim, metal ions mediating the interface, atom clashes in the interface etc. 
</p>
<p>
From the 3 votes a final decision is made ("Final" column). <font color="green">BIO</font> will mean that a majority
of the criteria indicated a biological interface and likewise for <font color="red">XTAL</font>.
This is the final prediction column and what you need to look at first.
</p>

<h2>
The main Interface table
</h2>
<p>
The other columns in the interface table correspond to a few important parameters that describe
the interfaces: the two chain codes of the partners (e.g. "A+B"), the Buried Surface Area upon 
formation (the sorting of the interfaces is based on this value), the crystal operator used to 
generate the second partner of the interface (e.g. "-X+1,Y-1/2,-Z") and the number of core residues in each side of 
the interface (e.g. "4+5")
</p>

<h2>
Viewing the interfaces in 3D
</h2>

<p>
<img class="eppic-iframe-right-img" src="resources/images/pse-withsurface-verysmall.png" 
 alt="3D view in PyMOL as it appears with the .pse files (surface representation was added)">
</img>

The thumbnails in the Interface table give a visual cartoon representation of the interfaces. By 
clicking on them one gets an interactive 3D view, depending on selected mode: as browser 
embedded Jmol viewer (no need for local viewer), as a PDB file downloadable to a local 
molecular viewer, or as PyMOL session file (.pse) to be opened in local PyMOL.
In both the jmol and the pse modes the two protomers are represented as cartoons with interface 
residues also shown as sticks. Core residues from both protomers are shown in two different shades 
of red.
</p> 
<p>
The pse mode is particularly useful as it is already prepared with selections of the different parts of the 
interface and nicely depicted. Additionally the entropy values for each residue are written as
b-factors in the .pdb and .pse files. Thus one can use the PyMOL "color by b-factor" functionality 
(spectrum command) to show the selection pressure pattern mapped on the 3-dimensional structure: blue colors
indicate low entropies and colors toward reds correspond to increasingly higher entropies. The surface
representation is particularly helpful here in order to see surface residues only.   
</p>





<h2>
The Residue Details table
</h2>
<p> 
Clicking the "Details" button in the last column will give you a detailed view of scores and surface 
areas per residue. In the residues table, only the residues at the interface are shown by default. To 
see all residues of the two chains one needs to switch the view with the drop down button on the upper 
right of the residues window. 
</p>


<h2>
Top panel
</h2>
<p>
The top panel contains additional information about the sequences and parameters used to run
the evolutionary predictions. On the top left corner there is a line for each of the distinct
protein sequences present in the protein structure. Clicking on the chain name one can see
the alignment to the reference UniProt sequence, where mismatching residues (normally engineered artifacts)
are shown in red. By clicking the download link on the bottom right corner one gets a single zip file 
containing the text output files with the results for all interfaces.
</p>


<h2>
A note on assemblies
</h2>
<p>
At the moment EPPIC classifies only pairwise interfaces in protein
crystals. It does not consider interfaces in the context of the full
biological assembly and does not group single interface calls into a
biological unit call. 
We intend to implement that in future versions of the software. 
In the meantime please keep in mind that some interfaces may be called <font color="red">XTAL</font> when they
are actually <font color="green">BIO</font> if considered in the context of the assembly: they are
"induced" by the topology or point-group symmetry of the assembly. 
</p>



<h2>
Job identifiers
</h2>
<p>
Every new job run on the server is assigned a unique job identifier, a long alpha numerical string 
that is only known by the user that submits it and that is very hard to guess. 
It is recommended that you give an email address while submitting so that you receive the URL with 
the job identifier in your inbox. Otherwise you will have to bookmark it or keep a record of it yourself.  
It is always possible to retrieve the job by using the URL http://www.eppic-web.org/ewui/#id/&lt;my_job_id&gt;.
Whether the job is still running or already done, the URL will show its current status and 
automatically display the final results whenever it is finished. You can even share the URL with 
colleagues.
</p>

<!-- <img src="resources/images/EPPIC-screenshot-3hbx-small.png"></img> -->


<p>
Please cite <a href="http://www.biomedcentral.com/1471-2105/13/334/abstract">Duarte et al., BMC Bioinformatics 2012</a> if you use this service for 
your research.
</p>

<p>
Do get in <a href="mailto:eppic@systemsx.ch">contact</a> with us if you want to report 
problems or send any kind of feedback.
</p>

<p>
With support from:
</p>

</div>

<div class="eppic-iframe-bottom-logos">
<a href="http://www.psi.ch">
<img class="eppic-iframe-left-img" src="resources/images/psi-logo.png"></img>
</a>

<a href="http://www.sybit.net">
<img class="eppic-iframe-right-img" src="resources/images/sybit-logo.png"></img>
</a>

</div>


<!--
<a href="http://www.snf.ch">
<img src="resources/images/SNSF_Logo.jpg"></img>
</a>
-->

<!--
<a href="http://www.ethz.ch">
<img class="eppic-iframe-bottom-img" src="resources/images/eth_logo.gif"></img>
</a>
-->
