<head>
<link rel="stylesheet" type="text/css" href="eppic-static.css">
<link href='http://fonts.googleapis.com/css?family=Open+Sans:400,700,400italic,700italic' rel='stylesheet' type='text/css'>
</head>

<div class="eppic-iframe">
	<div class="eppic-iframe-content">


<img class="eppic-iframe-top-img" src="resources/images/eppic-logo.png">


<h1>
Release log
</h1>

<h2>
2.0.4 - 17th January 2014
</h2>

<ul class="eppic-iframe-log-list">
<li>
Major user interface change:
	<ul>
		<li>Upgraded to GXT 3.0.1</li>
		<li>Introduced a top navigation bar for quick help, downloads and also to load new PDB structures</li>
		<li>Added information icons to pop-up more information when required</li>
		<li>Changes in "My Jobs" panel</li>
		<li>Improved Help page</li>
	</ul>
</li>
<li>Added a new homologs information window providing details on the homologs of each chain in the structure.</li>
<li>Introduced interface clustering. A radio button on top of the interfaces table now triggers a cluster view of similar interfaces.</li>
<li>Introduced a new FAQ page.</li>
<li>Introduced a new Publications page listing relevant references.</li>
</ul>

<h2>
2.0.3 - 27th November 2013
</h2>

<ul class="eppic-iframe-log-list">
<li>
Improvements in top panel: now split into experimental info panel and sequence info panel. 
Also some improvements in residue details window: scores are now better displayed.
</li>
<li>
Cleaned up the input page: only alignment parameters can now be modified in advanced input.
</li>
<li>
Issues with windows and tooltips displaying badly in IE are now gone thanks to GXT 2.3
</li>
<li>
Warnings for low resolution or high R-free added.
</li>
<li>
Interfaces below 35 Ã…<sup>2</sup> are not displayed anymore.
</li>
<li>
Fixed issue in interfaces calculation: the interface search extends now to more distant cells,
missing interfaces in structures such as 1was will now appear. 
</li>
<li>
A few bug fixes
</li>
</ul>


<h2>
2.0.2 - 22nd August 2013
</h2>

<ul class="eppic-iframe-log-list">
<li>
Precomputed EPPIC results are now available for the whole PDB, after typing a 
PDB code the results will appear immediately.
</li>
<li>
Now operators displayed as symmetry icons with the algebraic operators 
in tooltips. Operators colored in red to mark infinite interfaces.
</li>
<li>
A few bug fixes. In particular issue with crashes in parsing PDB files having no 
TER records (or badly placed ones) is now fixed.
</li>
<li>
The static pages are now fully crawlable by search engines.
</li>
<li>
Some style improvements
</li>
</ul>


<h2>
2.0.1 - 19th April 2013
</h2>

<ul class="eppic-iframe-log-list">
<li>
Fixed displaying problems in residue details table, the ugly white frames disappeared 
</li>
<li>
Now the homologs panel in top panel contains links to download the 
colored-by-entropies PSE files produced by the EPPIC command line 
program. Very useful to get a visual impression of the conservation of residues
at the surface of the structure. This shows graphically why EPPIC works
</li>
<li>
Better handling of UniProt reference mapping, especially for multi-segment mappings
</li>
<li>
More warnings added: unreliability of surface residues (surface residues not aligning 
to UniProt reference), if too many then NOPRED call is given
</li>
<li>
Display and style improvements
</li>
<li>
A few bug fixes
</li>
</ul>


<h2>
2.0.0 - 15th February 2013
</h2>

<ul class="eppic-iframe-log-list">
<li>
First full public release
</li>
<li>
Cofactors of more than 40 non-Hydrogen atoms are now taken into account. They will be considered as 
part of the protomers in ASA calculations. This affects especially predictions in small proteins 
with large cofactors like hemoglobin
</li>
<li>
PDB residue numbers are now used for warning messages and output jmol, pdb and pse files 
</li>
<li>
The missing eppic.conf file is now provided in downloadable eppic.zip
</li>
<li>
The guessing of file types is now more flexible, it will allow PDB files starting with:
HEADER, TITLE, COMPND, SOURCE, EXPDTA, REMARK, DBREF, SEQRES, CRYST1, MODEL, HELIX, SHEET. 
Any other file not starting with one of these lines will be rejected
</li>
<li>
Blast results are now precomputed for all PDB entries which makes runs for deposited PDBs a lot faster 
</li>
<li>
The web session will now expire after 24 hours of inactivity. The "My Jobs" list will then be lost
</li>
<li>
Style improvements 
</li>
<li>
Many bug fixes, especially in relation to parsing PDB and mmCIF files
</li>

</ul>




<h2>
1.9.6 - 22nd December 2012
</h2>

<ul class="eppic-iframe-log-list">
<li>
First public pre-release
</li>
</ul>



</div>
<div class="eppic-iframe-footer">
			<div class="eppic-iframe-footer-left">
				<p>
					Please cite <a href="http://www.biomedcentral.com/1471-2105/13/334"
						target="_blank"> Duarte JM, Srebniak A, Schaerer MA, Capitani
						G, BMC Bioinformatics 2012</a> if you use this service for your
					research.
				</p>

				<p>
					Do get in <a href="mailto:eppic@systemsx.ch">contact</a> with us if
					you want to report problems or send any kind of feedback. Follow
					updates on: <a href="https://twitter.com/eppicweb"
						class="twitter-follow-button" data-show-count="false">Follow
						@eppicweb</a>
					<script>!function(d,s,id){var js,fjs=d.getElementsByTagName(s)[0],p=/^http:/.test(d.location)?'http':'https';if(!d.getElementById(id)){js=d.createElement(s);js.id=id;js.src=p+'://platform.twitter.com/widgets.js';fjs.parentNode.insertBefore(js,fjs);}}(document, 'script', 'twitter-wjs');</script>
				</p>

			</div>
			<div class="eppic-iframe-footer-right">
				<p>With support from:</p>



				<!-- <div class="eppic-iframe-bottom-logos"> -->

				<p>
					<a href="http://www.psi.ch" target="_blank"> <img width="100px"
						src="resources/images/psi-logo.png"></img>
					</a> <a href="http://www.sybit.net" target="_blank"> <img
						width="100px" src="resources/images/sybit-logo.png"></img>
					</a> <a href="http://www.snf.ch" target="_blank"> <img
						width="100px" style="float: right;"
						src="resources/images/SNSF_Logo.jpg"></img>
					</a>
				</p>
			</div>
		</div>
</div>
