<head>
<link rel="stylesheet" type="text/css" href="eppic-static.css">
<link
	href='https://fonts.googleapis.com/css?family=Open+Sans:400,700,400italic,700italic'
	rel='stylesheet' type='text/css'>
</head>

<body>

	<script type="text/javascript">
		function reloadPage(url) {
			window.top.location=url;
		}
	</script>

	<div class="eppic-iframe-content">

		<img class="eppic-iframe-top-img"
			src="resources/images/eppic-logo.png">
		<div class="eppic-help-subtopic" id="about">

			<h1>The EPPIC server</h1>
			<div class="eppic-help-navigation">
				<div class="eppic-help-navigation-header">Navigation:</div>
				<ul>
					<li><a href="#input">Inputting data</a></li>
					<li><a href="#assemblies-table">The assemblies table</a></li>
					<li><a href="#interfaces-table">The interfaces table</a></li>
					<li><a href="#3D">Viewing interfaces and assemblies in 3D</a></li>
					<li><a href="#lattice">The lattice graph</a></li>
					<li><a href="#information-panel">Information panel</a></li>
					<li><a href="#homologs-window">The homologs information
							window</a></li>
					<li><a href="#residue-details">The Details window</a></li>
					<li><a href="#seq-clusters">Searching similar structures</a></li>
					<li><a href="#known-issues">Known issues</a></li>
					<li><a href="#misc">Miscellaneous</a></li>
					<li><a href="#funding">Funding</a></li>
				</ul>
			</div>

			<h4>EPPIC (Evolutionary Protein-Protein Interface Classifier)
				aims at predicting the quaternary structure of proteins from
				crystal structures. It first classifies the interfaces present in a crystal
				structure to determine whether they are biologically
				relevant or not. Then it enumerates
				all topologically valid assemblies in the crystal resulting from different
				combinations of the interfaces. Finally it provides
				a prediction of the most likely quaternary assembly based on the individual
				interfaces' scores.</h4>

			<p>In addition to that, it provides information of general use
				for a number of structural biology applications:</p>
			<ul>
				<li>Precomputed Multiple Sequence Alignments (MSAs) of closely
					related homologs (within 60% sequence identity)for every protein in
					the PDB or for user-uploaded structures</li>
				<li>Colored-by-sequence-entropy surface representation using the
					browser embedded <a href="https://github.com/arose/ngl">NGL</a> viewer. Very useful for looking at features present at protein
					surfaces: interfaces, binding sites etc. (<a
					href="#homologs-window">more</a>)
				</li>
				<li>Number and distribution of core residues in interfaces</li>
				<li>Symmetry features of interfaces: the crystallographic
					operators generating each interface are provided and are depicted
					in red if they are conducive to infinite assemblies. Also, one can
					compare the left- and right-hand side of the Details table of each
					interface to see if it is <a
					href="http://en.wikipedia.org/wiki/Heterologous" target="_blank">isologous</a>
					or <a href="http://en.wikipedia.org/wiki/Heterologous"
					target="_blank">heterologous</a>
				</li>
			</ul>
		</div>

		<!---------------------
		    Topics  start here
		 ----------------------->

		<!-- Input Data -->
		<h2 id="input">Inputting data</h2>
		<div class="eppic-help-subtopic">
			<div class="eppic-help-subtopic-right-small">
				<img class="eppic-help-subtopic-image-right-nozoom"
					src="resources/images/screenshots/input-parameters-upload-file.png"
					title="Screenshot of the input panel with advanced options"></img>
			</div>
			<div class="eppic-help-subtopic-left-big">
				<p>The server has precomputed results PDB-wide, if you are interested
					in a specific PDB id you can simply enter it and get results immediately.
					You can also submit your own protein structure by uploading a
					PDB/mmCIF file. If you upload your own
					file, the calculation will be triggered and the results are produced within
					a few minutes. Structures with many protein entities (unique sequences) will take
					longer.</p>

				<p>When you upload a file, the various input parameters can be changed via the Advanced
					tab in the input page. You have the following options:</p>
				<ul>
					<li><strong>Identity soft cut-off:</strong> The soft sequence
						identity cut-off for BLAST hits to be used in the alignment. If
						there are not enough sequences at this cut-off then the threshold
						is relaxed until the hard cut-off is reached.</li>
					<li><strong>Identity hard cut-off:</strong> The hard sequence
						identity cut-off for BLAST hits to be used in the alignment. If
						there are not enough sequences at soft cut-off then the threshold
						is relaxed until this hard cut-off.</li>
					<li><strong>Max num of sequences:</strong> The maximum number
						of sequences to be taken from the BLAST hits.</li>
					<li><strong>Search Mode:</strong> The PDB sequence is first
						matched to its UniProt entry and that sequence is then used for
						blasting. In Global mode the full length UniProt is used, in local
						only the matching PDB part of the UniProt sequence is used.</li>
				</ul>

				<p>
					Every new job run on the server is assigned a unique job
					identifier, a long alphanumerical string that is only known by the
					user who submits it and that is very hard to guess. This keeps the
					privacy of your data (<a href="#job-identifiers">more</a>).
				</p>
			</div>
		</div>

		<h2 id="assemblies-table">The assemblies table</h2>
		<p>All the topologically valid assemblies detected in the crystal are
		listed. Valid assemblies are those that exhibit Point Group Symmetry and
		that are isomorphous across the unit cell, i.e.
		all molecules are interacting in the same way throughout the lattice.
		</p>
		<div class="eppic-help-subtopic">
			<div class="eppic-help-subtopic-left-small">
				<img class="eppic-help-subtopic-image"
					src="resources/images/screenshots/assemblies-table.png"
					title="Screenshot of the assemblies table"></img>
			</div>
			<div class="eppic-help-subtopic-right-big">
				<p>
					Each assembly is represented by a thumbnail image in fat ribbon representation
					with different subunits in different colors (1st column). Also a 2-dimensional
					graph diagram of the assembly is shown, with nodes being the chains and edges the
					interfaces between them (2nd column). Distinct molecular entities and distinct
					interfaces are depicted with different colors. Next appear the macromolecular size,
					stoichiometry (the successive letters represent different molecular entities, not chain ids)
					and point group symmetry (Cn for cyclic symmetries, Dn for dihedral
					symmetries, T for tetrahedral, O for octahedral and I for icosahedral).
				</p>
				<p>
					Note that disjoint assemblies will be shown separated by commas in the size, symmetry and
					stoichiometry columns.
					Disjoint assemblies are those where not all components of the crystal form a single assembly but
					instead several disjoint ones, e.g. a crystal containing 2 protein entities will always contain
					a disjoint assembly formed by 2 independent monomers (one for each entity).
				</p>
				<p>
					The prediction column
					provides the predicted assembly (marked as <font color="green">BIO</font>), while
					all others are marked as <font color="red">XTAL</font>. The predicted assembly
					corresponds to that with the higuest calculated
					probability, based on the scores of the interfaces that form it. The probability values
					for each assembly appear in their own column too. An estimation of the
					prediction confidence is provided with star icons: golden star
					<img src="resources/icons/excellent.png" style="width: 20px; height: 20px"></img>
					for high confidence, gray star <img src="resources/icons/good.png" style="width: 20px; height: 20px"></img>
					for medium confidence, no star for low confidence.
				</p>
				<p>
					For PDB deposited
					entries, the assembly corresponding to the first PDB biounit assignment is marked with a small
					<img src="resources/icons/pdb_biounit_annotation.png"></img> icon.
					The last column shows how many interfaces compose the assembly, clicking on the link
					the interfaces table is shown with only those interfaces belonging to the assembly.
				</p>
				<p>
					One can switch over to the interfaces table by clicking on the
					"<strong>View All Interfaces</strong>"
					link provided on top of the table.
				</p>
			</div>
		</div>

		<!-- Main Interface table -->
		<h2 id="interfaces-table">The interfaces table</h2>
		<p>This table provides a view of all the pairwise protein-protein interfaces
			present in the crystal. One can switch between this view and the assemblies table
			view by clicking on the "<strong>View All Assemblies</strong>" link provided on top of the table.
			An assessment of the biological relevance of each pairwise interface is provided,
			based on a geometrical and on two evolutionary criteria:
		</p>
		<ul>
			<li><b>geometry</b>: number of core residues (at <b>95%
					burial</b>), indicating how good the packing in the interface is</li>
			<li><b>core-rim score</b>: ratio of evolutionary signal
				(sequence entropy) of core residues (at <b>70% burial</b>) versus
				rim residues</li>
			<li><b>core-surface score</b>: a z-score of
				sequence entropy of core residues (at <b>70% burial</b>) versus
				random samples of all surface residues</li>
		</ul>


		<div class="eppic-help-subtopic">
			<div class="eppic-help-subtopic-left-small">
				<img class="eppic-help-subtopic-image"
					src="resources/images/screenshots/interfaces-table.png"
					title="Screenshot of the main interface table"></img>
			</div>
			<div class="eppic-help-subtopic-right-big">

				<p>Each of these indicators have predefined
					score thresholds to produce one of the calls:</p>

				<ul>
					<li><font color="green">BIO</font>, the interface is
						biologically relevant</li>
					<li><font color="red">XTAL</font>, the interface is only a
						crystal lattice contact</li>
					<li>NOPRED, there is not enough information available to make
						a decision (usually not enough sequence data)</li>
				</ul>



				<p>
					You can see the actual scores for each of the indicators by
					hovering the mouse pointer over the bio/xtal/nopred labels.
					Warnings may appear on the right side of the interfaces table
					marked by the icon <img src="resources/icons/warning_icon.png"
						style="vertical-align: middle"></img>. By hovering the mouse over
					the icon, one can see different issues for which the prediction
					should be taken with care, e.g. engineered residues at the
					interface core or rim, metal ions mediating the interface, atom
					clashes in the interface etc.
				</p>
				<p>
					The three scores are used to calculate a final score and a probability
					of the interface being biologically relevant (1 being certainly biological,
					0 certainly crystal contact). The call and the probability appear in the
					"Final" column.
					<font	color="green">BIO</font> will mean that the probability is above 0.5
					and <font color="red">XTAL</font> that the probability is below 0.5.
					This is the final prediction column and what you need to look at
					first. An estimated confidence level for the prediction is depicted with stars,
					golden star <img src="resources/icons/excellent.png" style="width: 20px; height: 20px"></img>
					for high confidence, gray star <img src="resources/icons/good.png" style="width: 20px; height: 20px"></img>
					for medium confidence, no star for low confidence.
				</p>



				<p>The other columns in the interface table correspond to a few
					important parameters describing the interfaces: the two chain codes
					of the partners (e.g. "A+B"), the Buried Surface Area upon
					interface formation (interface sorting is based on this value), the
					icon of the crystal operator used to generate the second partner of
					the interface and the number of core residues on each side of the
					interface (e.g. "4+5").</p>
				<p>
					The operators are represented as icons to show at a glance what
					kind of crystallographic symmetry is present at the interface. The
					actual full algebraic operator (e.g. "-X+1,Y-1/2,-Z") can still be
					seen by hovering the mouse over the icon. The icons used for the
					operators are mostly the standard ones found in crystallographic
					tables: <img class="eppic-iframe-optype-img"
						src="resources/icons/optype_AU.png"></img> the identity operator
					(i.e. an interface in the asymmetric unit), <img
						class="eppic-iframe-optype-img"
						src="resources/icons/optype_XT.png"></img> a crystal translation
					(integer) without rotation, <img class="eppic-iframe-optype-img"
						src="resources/icons/optype_FT.png"></img> a re-centering
					translation without rotation, <img class="eppic-iframe-optype-img"
						src="resources/icons/optype_2.png"></img> a 2-fold axis, <img
						class="eppic-iframe-optype-img"
						src="resources/icons/optype_2S.png"></img> a 2-fold screw axis, <img
						class="eppic-iframe-optype-img" src="resources/icons/optype_3.png"></img>
					a 3-fold axis, <img class="eppic-iframe-optype-img"
						src="resources/icons/optype_3S.png"></img> a 3-fold screw axis, <img
						class="eppic-iframe-optype-img" src="resources/icons/optype_4.png"></img>
					a 4-fold axis, <img class="eppic-iframe-optype-img"
						src="resources/icons/optype_4S.png"></img> a 4-fold screw axis, <img
						class="eppic-iframe-optype-img" src="resources/icons/optype_6.png"></img>
					a 6-fold axis, <img class="eppic-iframe-optype-img"
						src="resources/icons/optype_6S.png"></img> a 6-fold screw axis.
				</p>
				<p>
					For the rare cases where a protein is crystallized in non-chiral
					space groups (e.g. racemic mixtures) there are additional
					operators: <img class="eppic-iframe-optype-img"
						src="resources/icons/optype_-1.png"></img> an inversion centre, <img
						class="eppic-iframe-optype-img"
						src="resources/icons/optype_-2.png"></img> a mirror plane, <img
						class="eppic-iframe-optype-img"
						src="resources/icons/optype_GL.png"></img> a glide plane, <img
						class="eppic-iframe-optype-img"
						src="resources/icons/optype_-3.png"></img> an improper 3-fold
					axis, <img class="eppic-iframe-optype-img"
						src="resources/icons/optype_-4.png"></img> an improper 4-fold
					axis, <img class="eppic-iframe-optype-img"
						src="resources/icons/optype_-6.png"></img> an improper 6-fold
					axis.
				</p>
				<p>Some of these operators lead to formation of infinite
					interfaces if occurring between two crystallographically-related
					copies of the same molecule (e.g. A+A). This happens for both the
					pure translations and for any of the screw rotations and it is
					generally a very strong indication of a crystal contact. In those
					cases we color the operator icon in red. The final call does note
					take that information into account, but this is very important
					for the enumeration of valid assemblies in the crystal shown in
					the assemblies table.</p>
			</div>
		</div>


		<h3>Clustering similar interfaces</h3>
		<div class="eppic-help-subtopic">
			<div class="eppic-help-subtopic-left-big">


				<p>Clusters of similar interfaces can be shown by checking the
					"Group similar interfaces" box. This ensures that interfaces
					that share a certain amount of contact similarity are grouped together. For
					each cluster, a header row displays a thumbnail representation of
					all member interfaces, while a bottom row reports the number
					of contributing interfaces and their average area. Additionally
					for each of the scores an average is calculated and a call
					produced based on these averaged scores. This should provide a better
					prediction than individual interface scores, since the possible
					differences between the chains are averaged out.</p>
			</div>
			<div class="eppic-help-subtopic-right-small">
				<img class="eppic-help-subtopic-image-right"
					src="resources/images/screenshots/interfaces-table-with-clusters.png"
					title="Screenshot of the main interface table with similar interfaces clustered together."></img>
			</div>
		</div>



		<h2 id="3D">Viewing interfaces and assemblies in 3D</h2>
		<div class="eppic-help-subtopic">
			<div class="eppic-help-subtopic-left-small">
				<img class="eppic-help-subtopic-image-nozoom"
					src="resources/images/screenshots/interface-view-ngl.png"
					title="3D view of an interface as displayed by the NGL viewer">
			</div>

			<div class="eppic-help-subtopic-right-big">
				<p>The thumbnails in the Assembly and Interface tables give a visual cartoon
					representation of the assemblies and interfaces. By clicking on them one gets an
					interactive 3D view, depending on the selected mode: as browser
					embedded <a href="https://github.com/arose/ngl">NGL</a> viewer (no need for local viewer) or as a mmCIF file
					downloadable to a local molecular viewer. In the interface view
					the two protomers are represented as cartoons with interface residues
					also shown as sticks. Core residues from both protomers are shown
					in two different shades of red. The sequence entropy values for each residue
					are written as b-factors in the .cif file. </p>
				<p>Once in the NGL window, by pressing "p" the user can get
					a surface representation of the subunits with sequence conservation
					values from the Multiple Sequence Alignment mapped onto it in a color code:
					blue colors indicate low entropies (high conservation) and colors toward yellow
					correspond to increasingly higher entropies (lower conservation). Pressing "n"
					the surface representation toggles through the chains.
					This representation is particularly useful in order to get a visual impression of
					the selection pressure patterns on the surface of a given
					chain. Residues that
					could not be aligned (for instance because they belong to an
					engineered tag) appear in red.
				</p>
				<h3>Changing 3D-Viewing mode:</h3>
				<p>The 3D-Viewer can be changed using the button "Change 3D
					Viewer" on the top-left corner of the main interface table.
					Clicking the button will open a drop-down menu where one
					can select any one of the two modes described above.</p>
			</div>
		</div>

		<h2 id="lattice">The lattice graph</h2>
		<div class="eppic-help-subtopic">
			<div class="eppic-help-subtopic-right-small">
				<img class="eppic-help-subtopic-image-right-nozoom"
					src="resources/images/screenshots/lattice-graph-3d.png"
					title="3D view of the lattice graph">
			</div>

			<div class="eppic-help-subtopic-left-big">
				<p>
					The assemblies are analysed through a graph representation of the chains
					and interfaces. This lattice graph is a periodic graph, with chains being
					the nodes and interfaces the edges. Visual tools to look at the graph are
					provided. They can be useful in understanding the crystal packing and
					the different possible assemblies that can be constructed with the
					given connectivity.
				</p>
				<ul>
					<li><strong>3D lattice graph view with <a href="https://github.com/arose/ngl">NGL</a></strong>:
						the graph is shown in 3D overlayed
						on top of the protein in semi-transparent cartoon represenatation.
						The nodes are placed at the centroids of the chains. Different colors
						are assigned to different molecular entities and to distinct interfaces.
						The view can be obtained by clicking on the "<strong>View Unit Cell</strong>" or
						"<strong>View Assembly in Unit Cell</strong>"
						links in the information panel.
					</li>
					<li><strong>2D lattice graph view (visjs)</strong>: the graph is shown in a dynamic 2D layout.
					Different colors
					are assigned to different molecular entities and to distinct interfaces. This view can be
					obtained by clicking on any of the assembly diagram thumbnails either on the assemblies
					table or on the assembly subpanel on top.
					</li>
				</ul>
			</div>
		</div>

		<h2 id="information-panel">Information panel
		</h2>
		<p>
			The top panel contains information about the assemblies and of the sequences and
			alignment used for the evolutionary scores.
			The information displayed depends on the context: in <strong>All Interfaces</strong>
			and <strong>All Assemblies</strong> views only 2 subpanels are shown with assemblies
			and sequence information (left screenshot below). The first subpanel on the left
			shows the number of assemblies, interfaces and interface
			clusters together with links to
			show each of the views. The second subpanel shows sequence and alignment information.
		</p>
		<p>
			In <strong>Single Assembly</strong> view, 3 subpanels are
			shown (right screenshot below). The first subpanel on the left
			contains information about the currently displayed assembly and its
			topology, depicted with the 2D graph diagram. A link is also provided to
			show the 3D lattice graph view.
			 The third subpanel is again the
			sequence/alignment information.
		</p>
		<img class="eppic-help-subtopic-image-no-animation"
			src="resources/images/screenshots/info-panel1.png"
			title="Screenshot of information panel"></img>
			<img class="eppic-help-subtopic-image-no-animation"
				src="resources/images/screenshots/info-panel2.png"
				title="Screenshot of information panel"></img>
		<p>
			The <b>Sequence Information</b> panel contains information about the
			sequences and their homologs. There is a line for each of the
			distinct protein sequences present in the protein structure. If too
			many chains are present, arrows will appear on the left of the panel
			to navigate to other chains. Each line displays some information about
			the homologs and some icons that lead to some actions for the given sequence
			entity:
		</p>
		<div class="eppic-help-subtopic">

			<div class="eppic-help-subtopic-right-small">
				<img class="eppic-help-subtopic-image-right-nozoom"
					src="resources/images/pse-color-entropies.png"
					title="Protein surface rendering with colors representing sequence entropies. This view can be obtained in the embedded NGL viewer by pressing p."></img>
			</div>

			<div class="eppic-help-subtopic-left-big">

				<ol>
					<li><b>Chain identifier(s)</b>: the first letter is the chain
						identifier of the representative sequence. This letter can be
						followed by other letters (in brackets). These are the chain
						identifiers of other identical sequences in the structure. Click
						on the <img class="eppic-iframe-optype-img" src="resources/icons/sequence_14.png"></img> icon to see
						the alignment of the chain to its reference
						UniProt sequence (based on <a
						href="http://www.ebi.ac.uk/pdbe/docs/sifts/" target="_blank">SIFTS</a>
						mapping or found by blasting). Mismatching residues, normally
						engineering artifacts, are shown in red in the alignment.
						A link to the reference UniProt record is provided on the top,
						as well as the organism to which the entry belongs.</li>
					<li><b>Number of homologs</b>: whenever fewer than 10 homologs
						are found you will see a NOPRED call in the evolutionary predictor
						columns (core-rim and core-surface), indicating that the MSA is
						too small for a proper decision to be reached. The more sequences
						in the alignment the more reliable the evolutionary predictions
						will be, so looking at this number is important. Click on the <img
						class="eppic-iframe-optype-img" src="resources/icons/homologs_14.png"></img> icon
						for more information about the homologs (<a
						href="#homologs-window">more</a>).</li>
					<li><b>Similar structures in PDB</b>: click on the
						<img class="eppic-iframe-optype-img" src="resources/icons/related_14.png"></img> icon to find
						other PDB entries that contain chains similar in sequence (down to 30%
						sequence identity) to this one (<a href="#seq-clusters">more</a>).

					</li>
				</ol>
			</div>
		</div>

		<h3>UniProt database version</h3>
		<p>
			The header of the <b>Sequence Information</b> panel shows which
			version of the UniProt database version is used to find homologs for
			the EPPIC multiple sequence alignments. A new UniProt database
			release appears every month containing more and more sequences, in
			fact the growth is exponential thanks to new sequencing technologies
			(growth statistics can be seen <a
				href="http://www.ebi.ac.uk/uniprot/TrEMBLstats/" target="_blank">here</a>).
			With more sequences we can obtain better predictions (see this <a
				href="http://www.biomedcentral.com/1471-2105/13/334/figure/F6"
				target="_blank">figure</a>), thus the growth of the UniProt sequence
			database has quite an important effect on the accuracy of our method.
			We will try to maintain the results as up-to-date as possible and to
			update the PDB-wide precomputed results every month (for every
			UniProt update). In any case if you use our results <strong>it
				is important to quote the UniProt database version used</strong>.
		</p>



		<h2 id="homologs-window">The Homolog Information window</h2>
		<div class="eppic-help-subtopic">
			<div class="eppic-help-subtopic-left-small">
				<img class="eppic-help-subtopic-image-nozoom"
					src="resources/images/screenshots/homologs-window.png"></img>
			</div>
			<div class="eppic-help-subtopic-right-big">
				<p>The homologs information window provides number, identity
					distribution and coverage of the sequence homologs in the MSA used
					for entropy calculation. This information is given for all unique
					chains in the structure. Coverage is depicted with colored segments
					within a rectangle corresponding to the query sequence length.
					Identity is color-coded in quantized bins, with green representing
					higher identity and red lower identity. The exact percentage of
					identity is also provided near the bar. The last two columns in the
					table provide the taxonomy of the homolog. More details of the
					sequence of a particular homolog can be found by clicking the
					UniProt link in the first column of the table. Clicking on the
					links provided on the Genus names will open their corresponding
					Wikipedia pages.
				<p>
				<p>
					In the header one can find information about the query sequence and
					the identity threshold used. Together with that, the clustering
					threshold for redundancy elimination is also indicated: sequences
					are clustered so that no pair of sequences are more similar than
					this value. The link in the top-right corner of the window enables
					you to download the MSA of all homolog sequences (FASTA format).
					You will need an alignment viewer like <a
						href="http://www.jalview.org/" target="_blank">Jalview</a> to have
					it nicely displayed.
				</p>

			</div>
		</div>

		<h2 id="residue-details">The Residue Detail window</h2>
		<div class="eppic-help-subtopic">
			<div class="eppic-help-subtopic-right-small">
				<img class="eppic-help-subtopic-image-right"
					src="resources/images/screenshots/residues-window.png"
					title="Screenshot of the residue details window">
			</div>

			<div class="eppic-help-subtopic-left-big">
				<p>Clicking on the "Details" button in the last column of the
					main interfaces table will open a window displaying
					interface-specific details. The window shows separated information
					for each side of the interface (the chain identifiers corresponding
					to the two sides are shown at the top). Below each chain
					identifier, score details are provided. Underneath those, a table
					(residue table) gives per-residue information.</p>
				<p>
					Score details are shown on top of the two grids for each side of the
					interface: number of core residues at 70% burial (i.e. the core
					residues used for calculating the evolutionary scores), number of
					rim residues (i.e. all other interface residues), core-rim
					score and core-surface score.
				</p>
				<p>By default, the two residue tables show only interface residues. To
					see all residues of the two chains one needs to switch the view in
					the drop down menu in the upper right corner of the window.</p>
				<p>The first two columns of the residue tables are the residue
					numbers: "Serial" for the numbering based on the SEQRES
					sequence from 1 to n, "PDB Serial" for the numbering as it appears
					on the ATOM lines of the PDB file. Four pieces of information are
					given for each residue: total accessible surface area (ASA), total
					buried surface area (BSA) upon interface formation, ratio of burial
					(BSA/ASA) upon interface formation (core/rim assignments are based
					on this) and sequence entropy value from the MSA of homologs to the
					query sequence. More information about the homologs and alignments
					can be found on the Top Panel.</p>
				<p>
					At the bottom of the window a legend shows the color coding used
					for the residue tables. Dark red denotes core residues at <b>95%
						burial</b>, which are used for the geometry prediction. Lighter red
					corresponds to core residues at <b>70% burial</b>, used for
					evolutionary core-rim and core-surface predictions. Light blue is
					used for all other interface residues, white for non-interface
					surface residues and grey for fully buried (protein interior)
					residues.
				</p>
			</div>
		</div>

		<h2 id="seq-clusters">Searching similar structures (sequence clusters)</h2>
		<div class="eppic-help-subtopic">
			<div class="eppic-help-subtopic-left-small">
				<img class="eppic-help-subtopic-image"
					src="resources/images/screenshots/seq-clusters.png"></img>
			</div>
			<div class="eppic-help-subtopic-right-big">
				<p>You can search for PDB structures that contain sequences similar to
				a given chain. Results are sorted from the structures with the most similar
				sequences (top) to structures containing more dissimilar sequences. The sequence
				identity level given in the first column is color-coded to facilitates the search.
				Crystallographic quality and crystal cell parameters are shown in separate columns.
				Structures can be sorted using any of these columns, which for instance can be useful
				to find the highest resolution structure within a set.
				<p>
				<p>
				Double clicking on a row or on the PDB code will open the analysis page
				for that structure.
				</p>
				<p>
				Note that we don't always recalculate the sequence clusters after every PDB update. Thus
				it is possible that for some chains there will be no button available if the clusters are
				not calculated for the particular sequence yet.
				</p>
			</div>
		</div>

		<h2 id="known-issues">Known issues</h2>

		<ul>
			<li>
			<b>MHC and antibody interfaces not correctly predicted</b>: due to the special nature
			of the MHC and antibodies sequences, the evolutionary criteria used by EPPIC do not hold
			for them. Thus the predictions for interfaces with at least one MHC or antibody molecule
			will often be incorrect.
			</li>
		</ul>


		<h2 id="misc">Miscellaneous</h2>

		<h3 id="job-identifiers">Job identifiers</h3>
		<p>
			Every new job run on the server is assigned a unique job identifier,
			a long alphanumerical string that is only known by the user that
			submits it and that is very hard to guess. It is recommended that you
			give an email address while submitting so that you receive the URL
			with the job identifier in your inbox. Otherwise you will have to
			bookmark it or keep a record of it yourself. It is always possible to
			retrieve the job by using the URL
			<code>http://www.eppic-web.org/ewui/#id/&lt;my_job_id&gt;</code>
			. Whether the job is still running or already done, the URL will show
			its current status and automatically display the final results
			whenever it is finished. To share the results of a job with
			colleagues just send them the corresponding URL. The jobs will be
			stored in our servers for 1 month and then deleted.
		</p>
		<p>The PDB-wide precomputed results can be accessed directly by
			using the permanent URLs:</p>
		<p>
			<code>http://www.eppic-web.org/ewui/#id/&lt;PDB_code&gt;</code>
		</p>

		<h3 id="low-resolution">Low resolution and high R-Free warnings</h3>
		<p>If the X-ray resolution of a PDB entry is lower than 3.5 Å, a
			low resolution warning will appear in red next to the structure title
			on top. Whenever this happens one should treat all results with
			caution, since our predictions depend on accurate side chain
			positioning to calculate burial and assign interface core residues. A
			warning also appears in case the R-Free factor is higher than 35%.</p>


		<h3 id="download-results">Downloading results in XML format</h3>
		<p>By clicking the "Download XML" link below the main title in the
			Top Panel, one gets an XML file containing all the information
			about interfaces and scoring (including the MSA) for the structure
			being analyzed. This is useful for further data processing. XML
			downloads of precomputed PDB entries can be scripted by using URLs like:
			<code>http://www.eppic-web.org/ewui/ewui/dataDownload?type=xml&amp;id=PDB_code</code>.
			See the <a href="http://www.eppic-web.org/ewui/#downloads" onclick="reloadPage('http://www.eppic-web.org/ewui/#downloads');">Downloads page</a> for more information.
		</p>


		<h2 id="funding">Funding</h2>
		<p>
			Funding to the project came initially from the
			<a href="https://www.psi.ch" target="_blank">Paul Scherrer Institute</a> (2010-2014) and later
			from the <a href="http://www.snf.ch/" target="_blank">Swiss National Science Foundation</a> (2013-2016).
			Since 2016 the <a href="http://www.rcsb.org/pages/about-us/index" target="_blank">RCSB Protein Data Bank</a>
			has supported the project and enabled its continuation.
			The RCSB PDB is funded by a grant (DBI-1338415) from the
			<a href="http://www.nsf.gov/" target="_blank">National Science Foundation</a>, the
			<a href="http://www.nih.gov/" target="_blank">National Institutes of Health</a>, and the
			<a href="http://science.energy.gov/" target="_blank">US Department of Energy</a>.
		</p>


	</div>
</body>
